File: java/core/src/mail/client/Initializer.java
/**
 * Author: Timothy Prepscius
 * License: GPLv3 Affero + keep my name in the code!
 */
package mail.client;

import core.callback.Callback;
import core.callback.CallbackEmpty;
import core.constants.ConstantsStorage;
import core.util.LogNull;
import core.util.LogOut;

public class Initializer extends Servent<Master>
{
	static LogNull log = new LogNull(Initializer.class);
	int numRemainingIndexCaches = 3;
	
	public Initializer ()
	{
	}
	
	public void start ()
	{
		EventDispatcher e = master.getEventPropagator();
		
		e.add(Events.Initialize_Start, this, new Callback() {

			@Override
			public void invoke(Object... arguments)
			{
				onInitializeStart();
			}
		});
		
		e.add(Events.Initialize_IndexedCacheAcquired, this, new Callback() {

			@Override
			public void invoke(Object... arguments)
			{
				onIndexedCacheAcquired();
			}
		});
		

		e.add(Events.Initialize_IndexedCacheLoadFailed, this, new Callback() {

			@Override
			public void invoke(Object... arguments)
			{
				onIndexedCacheLoadFailed();
			}
		});
		
		e.add(Events.Initialize_IndexedCacheLoadComplete, this, new Callback() {

			@Override
			public void invoke(Object... arguments)
			{
				onIndexedCacheLoadComplete();
			}
		});
		
		e.add(Events.Initialize_FolderLoadComplete, this, new Callback() {

			@Override
			public void invoke(Object... arguments)
			{
				onFolderLoadComplete();
			}
		});
		
		e.signal(Events.Initialize_Start, (Object[])null);
	}
	
	public void onInitializeStart()
	{
		log.debug("onInitializeStart");
		
		master.getStore().getConnector().ensureDirectories_(
			new String[] {
				ConstantsStorage.CACHE,
				ConstantsStorage.NEW_IN_JSON,
				ConstantsStorage.NEW_OUT_JSON
			}
		).invoke();

		master.getCacheManager().start();
	}

	public void onIndexedCacheLoadFailed ()
	{
		log.debug("onIndexedCacheLoadFailed");

		master.getCacheManager().firstRunInitialization();
		master.getIndexer().firstRunInitialization();
		
		master.getEventPropagator().signal(Events.FirstRunInitialization, (Object[])null);
	}
	
	public void onIndexedCacheAcquired ()
	{
		log.debug("onIndexedCacheLoadAcquire");
		master.getCacheManager().deserializeIndexCaches();
	}

	public void onIndexedCacheLoadComplete ()
	{
		log.debug("onIndexedCacheLoadComplete");
		master.getIndexer().start();
	}
	
	public void onFolderLoadComplete()
	{
		log.debug("onFolderLoadComplete");

		master.getEventPropagator().signal(Events.Initialized, (Object[])null);
	}
}

File: cpp/core/mailiverse/mail/manager/Initializer.cpp
/**
 * Author: Timothy Prepscius
 * License: BSD + keep my name in the code!
 */

#include "Initializer.h"
#include "Master.h"
#include "Events.h"
#include "../model/Settings.h"
#include "mailiverse/External.h"
#include "mailiverse/core/constants/ConstantsSettings.h"
#include "mailiverse/core/constants/ConstantsPushNotifications.h"

using namespace mailiverse::mail::manager;
using namespace mailiverse::core::constants;

Initializer::Initializer() :
	ready(false)
{
}

Initializer::~Initializer()
{
}

bool Initializer::isReady ()
{
	return ready;
}

void Initializer::start ()
{
	getMaster()->getEventPropagator()->add(
		Events::Initialize_FirstRun, this,
		EventPropagator::Callback (
			utilities::newbindC_GV(this, &Initializer::onFirstRun)
		)
	);

	getMaster()->getEventPropagator()->add(
		Events::Initialize_IndexedCacheLoadComplete, this,
		EventPropagator::Callback (
			utilities::newbindC_GV(this, &Initializer::onCacheInitialized)
		)
	);

	getMaster()->getEventPropagator()->add(
		Events::Initialize_IndexedCacheLoadFailed, this,
		EventPropagator::Callback (
			utilities::newbindC_GV(this, &Initializer::onCacheFailed)
		)
	);

	getMaster()->getEventPropagator()->add(
		Events::Initialize_FolderLoadComplete, this,
		EventPropagator::Callback (
			utilities::newbindC_GV(this, &Initializer::onFolderLoadComplete)
		)
	);

	getMaster()->getEventPropagator()->add(
		Events::Initialized, this,
		EventPropagator::Callback (
			utilities::newbindC_GV(this, &Initializer::onInitialized)
		)
	);

	getMaster()->getEventPropagator()->add(
		Events::SettingsLoaded, this,
		EventPropagator::Callback (
			utilities::newbindC_GV(this, &Initializer::onSettingsLoaded)
		)
	);

	getMaster()->getCacheManager()->initialize();
}

void Initializer::onFirstRun ()
{
	getMaster()->getCacheManager()->onFirstRun();
	getMaster()->getIndexer()->onFirstRun();

	ready = true;
	getMaster()->getEventPropagator()->signal(Events::Initialized, NULL);
}

void Initializer::onCacheInitialized ()
{
	getMaster()->getIndexer()->initialize();

	ready = true;	
	getMaster()->getEventPropagator()->signal(Events::Initialized, NULL);
}

void Initializer::onCacheFailed ()
{

}

void Initializer::onFolderLoadComplete ()
{

}

void Initializer::onInitialized()
{
	getMaster()->getActions()->onInitialized();
}

void Initializer::onSettingsLoaded ()
{
	model::SettingsPtr settings = getMaster()->getCacheManager()->getSettings();
	
	bool hasNotifications = 
		settings->get(ConstantsSettings::NOTIFICATION_TYPE,ConstantsPushNotifications::NOTIFICATION_TYPE_NONE) !=
		ConstantsPushNotifications::NOTIFICATION_TYPE_NONE;
		
	if (hasNotifications)
		__onEnableDeviceNotifications__(true);
}